src/choose_formatter.cpp

```diff
@@ -28,233 +29,190 @@
-    using qi::lexeme;
-    using qi::raw;
-
-    // RFC2616 definition of a token
-    token %= +(char_(33, 126) -
-               (char_("(") | char_(")") | char_("<") | char_(">") | char_("@") |
-                char_(",") | char_(";") | char_(":") | char_("\\") |
-                char_("\"") | char_("/") | char_("[") | char_("]") |
-                char_("?") | char_("=") | char_("{") | char_("}")));
-    // RFC2616 definition of a quoted string
-    quoted_string %= lit("\"") >>
-                     *((char_(32, 126) - char_("\"")) | (lit("\\") >> char_)) >>
-                     lit("\"");
-
-    // the string here can be either a token/token pair (where token can include
-    // '*') or a single '*' character.
-    // TODO: this will accept '*/something', but that is an invalid mime type
-    // and should be rejected.
-    mime_type %= raw[(token >> lit("/") >> token) | lit("*")];
-    param %= token >> '=' >> (token | quoted_string);
-    range %= mime_type >> *(';' >> param);
-    start %= range % ',';
-  }
-
-  qi::rule<iterator, string()> token, quoted_string, mime_type;
-  qi::rule<iterator, pair<string, string>(), ascii::blank_type> param;
-  qi::rule<iterator, media_range(), ascii::blank_type> range;
-  qi::rule<iterator, std::vector<media_range>(), ascii::blank_type> start;
-};
-/*
-      = lit("* / *")      [_val = mime::any_type]
-      | lit("text/xml") [_val = mime::type::application_xml]
-      | lit("application/xml") [_val = mime::type::application_xml]
-#if HAVE_YAJL
-      | lit("application/json")[_val = mime::type::application_json]
-#endif
-      ;
-*/
-
-acceptable_types::acceptable_types(const std::string &accept_header) {
-  using boost::spirit::ascii::blank;
-  using iterator_type = std::string::const_iterator;
-  using grammar = http_accept_grammar<iterator_type>;
-
-  std::vector<media_range> ranges;
-  grammar g;
-  iterator_type itr = accept_header.begin();
-  iterator_type end = accept_header.end();
-  bool status = phrase_parse(itr, end, g, blank, ranges);
-
-  if (status && (itr == end)) {
-    for (media_range range : ranges) {
-      // figure out the mime::type from the string.
-      mime::type mime_type = mime::parse_from(range.mime_type);
-      if (mime_type == mime::type::unspecified_type) {
-        // if it's unknown then skip this type...
-        continue;
-      }
-
-      // figure out the quality
-      auto q_itr = range.params.find("q");
-      // default quality parameter is 1
-      float quality = 1.0;
-      if (q_itr != range.params.end()) {
-        quality = std::stof(q_itr->second);
-      }
-
-      mapping.insert(make_pair(mime_type, quality));
-    }
+  for (const auto &acceptedType : acceptedTypes)
+    mapping[acceptedType.mimeType] = acceptedType.q;
 
-  } else {
-    logger::message(fmt::format("Failed to parse accept header '{}'",
-                    accept_header));
+  if (mapping.empty()) {
     throw http::bad_request("Accept header could not be parsed.");
```

Comment:
This should be a `406 Not Acceptable` instead of a `400 Bad Request` if there are no accepted media types in the header. (Non-accepted types are currently simply skipped in `parse()`, which can cause the map to be empty)

Not sure if the constructor should be throwing exceptions at all though unless the input string is malformed?